### ISSUE TITLE
Add webhook for APIs defaulting and validation

### ISSUE THREAD

**Issue by ahg-g** [kubernetes-sigs/kueue#171]
A webhook for APIs defaulting and validation; see example: https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html

**Comment by knight42** [comment]
@alculquicondor Hi! Thanks for your comments, that would definitely be a good start!

I am trying to follow the steps described in https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html to scaffold the webhook, but I failed:
```
$ git diff PROJECT
diff --git PROJECT PROJECT
index bab428b..93389ef 100644
--- PROJECT
+++ PROJECT
@@ -12,7 +12,7 @@ resources:
   domain: x-k8s.io
   group: kueue
   kind: Queue
-  path: sigs.k8s.io/kueue/api/v1alpha1
+  path: sigs.k8s.io/kueue/apis/core/v1alpha1
   version: v1alpha1
 - api:
     crdVersion: v1

$ kubebuilder create webhook --group kueue --version v1alpha1 --kind Queue --defaulting --programmatic-validation
Writing kustomize manifests for you to edit...
Error: failed to create webhook: unable to scaffold with "kustomize.common.kubebuilder.io/v1": error updating resource: unable to update Resource (Path "sigs.k8s.io/kueue/apis/core/v1alpha1") with another with non-matching Path "sigs.k8s.io/kueue/apis/kueue/v1alpha1"
...
```
It seems the current project layout confused `kubebuilder`, shall we rename `apis/core` to `apis/kueue`? Or am I missing something?

**Comment by alculquicondor** [comment]
It seems that we didn't do a good job keeping the `PROJECT` file in-sync with the renames. It looks like you have to change:

```
group: core
```

**Comment by ahg-g** [comment]
Changing to the group to `core` is going to produce `core.x-k8s.io` as the group name for the apis instead of `kueue.x-k8s.io`; so I don't think we should change the group.

perhaps only update the paths.

**Comment by ahg-g** [comment]
I think for kubebuilder tool to work smoothly we indeed need to move `core` to `kueue`, which sounds redundant, but might unfortunately be necessary if we want to continue to rely on the kubebuilder tool

**Comment by alculquicondor** [comment]
Opened #219

**Comment by ahg-g** [comment]
Now that we started with `Workload`, do we believe we are done with defaulting and validation for that resource? Should some fields be marked immutable (e.g., PodSets)?

**Comment by alculquicondor** [comment]
as a hole? Are we sold on the idea of not adding the node affinities to the Workload, but add them directly to the Job?

**Comment by alculquicondor** [comment]
I'm trying to think of ways to avoid the custom controllers to look at the ResourceFlavors and only look at the Workload.

Although, just directly modifying the Workload's pod specs is probably not the best idea. Maybe at admission time we can add the node labels somewhere in the `.spec.admission` struct.

So, you are right, making the PodSpecs immutable makes sense.

**Comment by ahg-g** [comment]
Adding the labels to spec.admission makes sense I think.

ok, lets round up the workload validation with making the spec immutable and then perhaps we focus on the next one, queue is probably the easier?

@knight42 would you like to send a PR validating that the PodSet is immutable?

**Comment by knight42** [comment]
> @knight42 would you like to send a PR validating that the PodSet is immutable?

Yeah I am glad to do that!

**Comment by alculquicondor** [comment]
Did you already add a webhook for Queue? It would be useful to make `.spec.clusterQueue` immutable.

**Comment by knight42** [comment]
> Did you already add a webhook for Queue? It would be useful to make `.spec.clusterQueue` immutable.

Nope, now there is only webhook for `Workload`. I will add it in upcoming days.

**Comment by alculquicondor** [comment]
We should also validate Workload updates. We probably want to make most of the fields immutable. Also, we should disallow changing the queue of an Admitted workload.

**Comment by ahg-g** [comment]
To track what is left, lets try and list the validations we want to do:

**Workload validation**
1. Queue and Admitted fields are immutable on update once set. 

**Queue validation** 
Nothing to validate for now?

**ClusterQueue validation**
1. Two resources can either have completely different flavors or they must share all flavors and their order. 
2. Resource name is a valid name
3. NamespaceSelector is a valid selector
4. Validate quota (max > min)
5. Resource flavor name is a valid name

**ResourceFlavor validation**
1. [validate labels](https://github.com/kubernetes/kubernetes/blob/894a8fd4dd7fe40be5c035c03ce40c088c775521/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/validation/validation.go#L71)

**Comment by ahg-g** [comment]
@alculquicondor  anything else?

**Comment by alculquicondor** [comment]
edited the comment above

**Comment by ahg-g** [comment]
@kerthcet do you want to send a PR for CQ validation?

**Comment by kerthcet** [comment]
Yes, I'll finish it soon.

**Comment by ahg-g** [comment]
Almost there, only one left is validating CQ

**Comment by kerthcet** [comment]
FYI: We only leave one validation now.

`Two resources can either have completely different flavors or they must share all flavors and their order.`

**Comment by ahg-g** [comment]
> FYI: We only leave one validation now.
> 
> `Two resources can either have completely different flavors or they must share all flavors and their order.`

We need to do that as well, we have scheduling logic assuming that this is validated.

**Comment by ahg-g** [comment]
Done
